### PR TITLE
Added support for P8 certificate apple push notifications

### PR DESCRIPTION
**Pull request by mariannegru**
Apple push notifications using authentication tokens

http://help.apple.com/xcode/mac/current/#/dev54d690a66

### CODE REVIEW

File: lib/generators/templates/create_rapns_apps.rb [hunk 1]
```diff
@@ -14,3 +14,4 @@ def self.down
     drop_table :rapns_apps
   end
 end
+x
```
Comment (aried3r) [inline lib/generators/templates/create_rapns_apps.rb]: I guess this was mistakenly added here?

File: README.md [hunk 2]
```diff
@@ -79,6 +79,28 @@ You should also implement the [ssl_certificate_will_expire](https://github.com/r
 
 To use the newer APNs Api replace `Rpush::Apns::App` with `Rpush::Apns2::App`.
 
+To use the p8 APNs Api replace `Rpush::Apns::App` with `Rpush::Apnsp8::App`.
+
+```ruby
+app = Rpush::Apnsp8::App.new
+app.name = "ios_app"
+app.apn_key = File.read("/path/to/sandbox.pem")
```
Comment (aried3r) [inline README.md]: Isn't it a `.p8` file?

**Comment by aried3r** [comment]
Thanks for your PR! Sadly I cannot test this myself as I don't use APNS myself, but I'll try to have a closer look soon.

**Comment by soulfly** [comment]
@mariannegru  could you please provide more information in description for this PR, e.g. Apple links to what is P8 and how to generate it. Will be useful for testing. thanks!

**Comment by sveredyuk** [comment]
I can test this, because I need APNS token support

**Comment by sveredyuk** [comment]
@mariannegru Good job. I am testing it right now. Got `Errno::ECONNRESET, Connection reset by peer`. It's because of wrong credentials ?

**Comment by mariannegru** [comment]
@sveredyuk Can you check in sql if the apn_key is set? Must be something with `-----BEGIN PRIVATE KEY-----` and `-----END PRIVATE KEY-----`

I use fastlane to generate p8 files.

**Comment by sveredyuk** [comment]
Yes, apn_key is present and valid (i got it from apple dev site), JWT token is generating but response `Errno::ECONNRESET, Connection reset by peer`

**Comment by mariannegru** [comment]
Are you testing development or production? Maybe the device token is not for the right environment.

**Comment by sveredyuk** [comment]
for production, we are using production env even on staging devices

**Comment by mariannegru** [comment]
If you run the app from xcode I think you should use the development environment and make sure you have the rpush app configured for development.
Seams the Errno::ECONNRESET is retuned by inactivity or when you send device tokens to different environments.

We have it for 46 apps in production and is working fine.

**Comment by lenart** [comment]
Hey @mariannegru. What's holding this PR from being merged into master?

**Comment by aried3r** [comment]
I think it should be rebased once more to see if the issues with CI were fixed in `master`. Then I'd be open to merging this as an experimental feature. I have no way of testing this myself.

I'd rather have more people try out this feature as "experimental" than rely on people manually trying out this branch.

**Comment by sveredyuk** [comment]
We are waiting for this MR to be merged. I could test it again

**Comment by mariannegru** [comment]
@sveredyuk did you tested this again?

**Comment by sveredyuk** [comment]
@mariannegru just planed to test on weekend.

**Comment by aried3r** [comment]
Hey! Sorry for the long silence. I am currently preparing a 3.1.0 release and will try to tackle this next for a 3.2.0.

Unfortunately there are now some (expected) conflicts. If I find the time I'll try to resolve them myself, but I'd appreciate any help. One thing that would have to be changed is the name of the migration (`rpush_3_2_0_add_apns_p8` instead of just the `add_apns_p8`).

**Comment by mariannegru** [comment]
@aried3r Sorry for the delay, I've merged the latest and renamed to migrations to the next release

**Comment by aried3r** [comment]
Awesome! Thank you, I'll have another look and hope to merge it soon when I get the time.

**Pull request merged**